### PR TITLE
support query encoding in http, http, ftp and fix path decode issues

### PR DESCRIPTION
**Pull request by jrieken**
This PR is for https://github.com/microsoft/vscode/issues/25852 _and_ https://github.com/microsoft/vscode/issues/45515

### CODE REVIEW

**Comment by jrieken** [comment]
@isidorn these changes make some tests in debug land fail: https://github.com/microsoft/vscode/blob/c653508f853b6f2f7ac2611f3780b5106179aaaa/src/vs/workbench/contrib/debug/test/electron-browser/debugModel.test.ts#L393. It seems like the [debug source(https://github.com/microsoft/vscode/blob/c653508f853b6f2f7ac2611f3780b5106179aaaa/src/vs/workbench/contrib/debug/common/debugSource.ts#L52) is build on the (bug) behaviour described in https://github.com/microsoft/vscode/issues/45515, e.g when the path component contains an encoded / (`%2f`) it should not be decoded but kept as is

**Comment by isidorn** [comment]
Looking at it I am not sure why do we encode the path component [here](https://github.com/Microsoft/vscode/blob/8cccedb38fd282d9aa1437f26be4bdd0a581783e/src/vs/workbench/contrib/debug/common/debugSource.ts#L52)

@weinand introduced the encoding of the uri for the debug source so he would be better to comment here.

**Comment by jrieken** [comment]
I the goal is to create new uris from existing components that you should use [`URI.from`](https://github.com/microsoft/vscode/blob/31393cc31f63614a2d59e60c8d0be2d236038367/src/vs/base/common/uri.ts#L345)

**Comment by isidorn** [comment]
Yeah that seems like the way to go.
I can do the changes if @weinand agrees

**Comment by weinand** [comment]
@isidorn yes, please adapt.

**Comment by jrieken** [comment]
Great - let me know if you need help

**Comment by isidorn** [comment]
@jrieken thanks, it would be great if you could reveiw this PR I just created.https://github.com/microsoft/vscode/pull/74781

File: src/vs/base/common/uri.ts [hunk 1]
```diff
@@ -96,6 +96,14 @@ const _empty = '';
...
 const _regexp = /^(([^:/?#]+?):)?(\/\/([^/?#]*))?([^?#]*)(\?([^#]*))?(#(.*))?/;
 
+const _queryStringSchemes = new Set<string>();
+_queryStringSchemes.add('http');
+_queryStringSchemes.add('https');
+_queryStringSchemes.add('ftp');
+_queryStringSchemes.add('HTTP');
+_queryStringSchemes.add('HTTPS');
+_queryStringSchemes.add('FTP');
+
```
Comment (jrieken) [inline src/vs/base/common/uri.ts]: what about `hTtp`?

Comment (jrieken) [reply src/vs/base/common/uri.ts]: or `ftP`?

Comment (jrieken) [reply src/vs/base/common/uri.ts]: or `ftP`?

**Pull request merged**